rules/expiring-todo-comments.js

```diff
@@ -0,0 +1,390 @@
+			uses++;
+			const [{condition, version}] = packageVersions;
+
+			const pkgVersion = tryToCoerceVersion(pkg.version);
+			const desidedPkgVersion = tryToCoerceVersion(version);
+
+			const compare = semverComparisonForOperator(condition);
+
+			if (compare(pkgVersion, desidedPkgVersion)) {
+				context.report({
+					node: null,
+					loc: comment.loc,
+					messageId: MESSAGE_ID_REACHED_PACKAGE_VERSION,
+					data: {
+						comparison: `${condition} ${version}`
+					}
+				});
+			}
+		}
+
+		// Inclusion: 'in', 'out'
+		// Comparison: '>', '>='
+		for (const dependency of dependencies) {
+			uses++;
+			const targetPackageRawVersion = pkgDependencies[dependency.name];
+			const hasTargetPackage = Boolean(targetPackageRawVersion);
+
+			const isInclusion = ['in', 'out'].includes(dependency.condition);
+			if (isInclusion) {
+				const [trigger, messageId] =
+					dependency.condition === 'in' ?
+						[hasTargetPackage, MESSAGE_ID_HAVE_PACKAGE] :
+						[!hasTargetPackage, MESSAGE_ID_DONT_HAVE_PACKAGE];
+
+				if (trigger) {
+					context.report({
+						node: null,
+						loc: comment.loc,
+						messageId,
+						data: {
+							package: dependency.name
+						}
+					});
+				}
+
+				continue;
+			}
+
+			const todoVersion = tryToCoerceVersion(dependency.version);
+			const targetPackageVersion = tryToCoerceVersion(targetPackageRawVersion);
+
+			if (!hasTargetPackage || !targetPackageVersion) {
+				// Can't compare ¯\_(ツ)_/¯
+				continue;
+			}
+
+			const compare = semverComparisonForOperator(dependency.condition);
+
+			if (compare(targetPackageVersion, todoVersion)) {
+				context.report({
+					node: null,
+					loc: comment.loc,
+					messageId: MESSAGE_ID_VERSION_MATCHES,
+					data: {
+						comparison: `${dependency.name} ${dependency.condition} ${
+							dependency.version
+						}`
+					}
+				});
+			}
+		}
+
+		const pkgEngines = pkg.engines || {};
+
+		for (const engine of engines) {
+			uses++;
+			const targetPackageRawEngineVersion = pkgEngines[engine.name];
+			const hasTargetEngine = Boolean(targetPackageRawEngineVersion);
+
+			if (!hasTargetEngine) {
+				continue;
+			}
+
+			const todoEngine = tryToCoerceVersion(engine.version);
+			const targetPackageEngineVersion = tryToCoerceVersion(
+				targetPackageRawEngineVersion
+			);
+
+			const compare = semverComparisonForOperator(engine.condition);
+
+			if (compare(targetPackageEngineVersion, todoEngine)) {
+				context.report({
+					node: null,
+					loc: comment.loc,
+					messageId: MESSAGE_ID_ENGINE_MATCHES,
+					data: {
+						comparison: `${engine.name} ${engine.condition} ${engine.version}`
+					}
+				});
+			}
+		}
+
+		return uses === 0;
+	}
+
+	return {
+		Program() {
+			rules.Program(); // eslint-disable-line new-cap
+		}
+	};
+};
+
+const schema = [
+	{
+		type: 'object',
+		properties: {
+			terms: {
+				type: 'array',
+				items: {
+					type: 'string'
+				}
+			},
+
+			ignoreDatesOnPR: {
+				type: 'boolean'
+			}
+		}
+	}
+];
```

Comment:
Set the default values in the schema here instead of in the code.

Reply:
I'm kinda lost on this one, can you give me an example? Currently in this repo I found:

I'm doing like this:
```js
// catch-error-name.js
const create = context => {
	const options = {
		name: 'error',
		caughtErrorsIgnorePattern: '^_$',
		...context.options[0]
	};
```

```js
// filename-case.js
const create = context => {
	const chosenCases = getChosenCases(context);

// ...
function getChosenCases(context) {
	const option = context.options[0] || {};
```

```js
// no-keyword-prefix.js
const create = context => {
	const options = prepareOptions(context.options[0]);
```

```js
// prefer-add-event-listener.js
const create = context => {
	const options = context.options[0] || {};
```

This method sounds better:
```js
// prevent-abbreviations.js
const create = context => {
	const options = prepareOptions(context.options[0]);
```

After going to the site os JSON Schema I got into [this example with default](https://json-schema.org/understanding-json-schema/structuring.html?highlight=default#using-id-with-ref) I got this:

```js

const schema = [
	{
		type: 'object',
		properties: {
			terms: {
				type: 'array',
				default: ['todo', 'fixme', 'xxx'],
				items: {
					type: 'string'
				}
			},
			ignoreDatesOnPullRequests: {
				type: 'boolean',
				default: true
			}
		},
	}
];
```

But doesn't feel right since tests get `[]` as ctx.options and fail unless I handle on code. Any tip?

Reply:
I've also seen issues with `default` during test runs. I haven't dug into it yet but it's possible that the schema resolution is being skipped during tests.

Reply:
We should upgrade to ESLint 6, maybe we're lucky and it's been fixed already: https://github.com/sindresorhus/eslint-plugin-unicorn/issues/326

Reply:
I'm trying once again now that it's fixed. Already rebased. Will post result soon.

Reply:
Added schema defaults except for array. [Apparently they don't use `default` on array on eslint](https://github.com/eslint/eslint/blob/d058a8639a2ef8837974ed2b516c7ce8a79418d6/lib/rules/new-cap.js)